docs_src/design/adr/0023-North-South-Messaging.md

```diff
@@ -0,0 +1,527 @@
+
+The `envelope` will also contain the API version (something provided in the HTTP path when using REST).
+
+Command requests in HTTP may also contain ds-pushevent and ds-returnevent query parameters (for GET commands).  These will be optionally provided key/value pairs represented in the message `envelope`'s query parameters (and optionally allows for other parameters in the future).
+
+``` JSON
+{
+    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
+    "API":"V2",
+    "queryParams": {
+        "ds-pushevent":"yes",
+        "ds-returnevent":"yes",
+     }
+     ...
+}
+```
+
+#### Command Message Payload
+
+The **request** message `payload` to the command service and those relayed to the device service would mimic their HTTP/REST request body alternatives.  The `payload` provides details needed in executing the command at the south side.
+
+In the example GET and PUT messages below, note the `envelope` wraps or encases the message `payload`.  The payload may be empty (as is typical of GET requests).
+
+``` JSON
+{
+    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
+    "apiVersion": "v2",
+    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
+    "queryParams": {
+        "ds-pushevent":"yes",
+        "ds-returnevent":"yes",
+     }
+    "payload": 
+    {
+    }
+}
+
+{
+    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
+    "apiVersion": "v2",
+    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
+    "payload": 
+    {
+        "AHU-TargetTemperature": "28.5",
+        "AHU-TargetBand": "4.0",
+        "AHU-TargetHumidity": {
+            "Accuracy": "0.2-0.3% RH",
+            "Value": 59
+        }
+    }
+}
+```
+
+The **response** message `payload` would contain the response from the south side, which is typically EdgeX event/reading objects (in the case of GET requests) but would also include any error message details.
+
+Example response messages for a GET and PUT request are shown below.  Again, note that the message `envelope` wraps the response `payload`.
+
+``` JSON
+{
+    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
+    "apiVersion": "v2",
+    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
+    "errorCode": 0,
+    "payload": 
+    {
+    "event": {
+        "apiVersion": "v2",
+        "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
+        "deviceName": "string",
+        "profileName": "string",
+        "created": 0,
+        "origin": 0,
+        "readings": [
+            "string"
+            ],
+        "tags": {
+            "Gateway-id": "HoustonStore-000123",
+            "Latitude": "29.630771",
+            "Longitude": "-95.377603"
+            }
+        }
+    }
+}
+
+{
+    "Correlation-ID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
+    "apiVersion": "v2",
+    "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
+    "errorCode": 1,
+    "payload": 
+    {
+        "message": "string"
+    }
+}
+
+```
```

Comment:
Note, the use of pre-formatted text makes it a bit of a pita to review due to the horizontal scrolling involved. Just a suggestion to maybe not use this in future ADRs.

Reply:
noted